horovod/tensorflow/gradient_aggregation_eager.py

```diff
@@ -119,8 +119,8 @@ def _allreduce_helper(self, grads, vars):
     def _clear_vars(self):
         self.counter.assign(0)
         for idx in self.locally_aggregated_grads.keys():
-            self.locally_aggregated_grads[idx].assign_add(
-                -1 * self.locally_aggregated_grads[idx])
```

Comment:
Is this a breaking change for existing user code?

Reply:
inf - inf is NaN, so semantically it would definitely be an improvement to assign zero here.

However, I suspect that `tf.assign_add()` has been used here originally to avoid an intermediate extra memory allocation for the result of `tf.zeros_like()`. In the past I've seen a similar effect actually cause perceivable memory waste.

I wonder if this is still the case or if recent releases of TensorFlow can optimize `x.assign(tf.zeros_like(x))` appropriately..

Reply:
@tgaddair Is there a reason of introducing `tf.assign_add()` originally?

Reply:
@maxhgerlach i had the same doubt initially, but according to [this](https://www.tensorflow.org/guide/tensor), it seems like even the old assign_add created a new buffer. I guess ultimately the proper way is to use the [in-place](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/inplace-update) c++ api which is a bit involved than the scope of the pr.

Reply:
I would assume:
```
a.assign_add(-1 * a)
```
consumes the same amount of memory as:
```
a.assign(tf.zeros_like(a))
```
since both likely create temporaries, as mentioned by @Tixxx.

Is there a TF API to just set all values of a tensor to a scalar value?

Reply:
We should go into [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/assign_op.h#L90-L104) route when assigning zero to local gradient. What happens afterwards will be the same for [Assign](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/dense_update_ops.cc#L39-L40) and [AssignAdd](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/dense_update_ops.cc#L80-L82), the only difference is for assign the update is `params.device(d) = update;` and for assignAdd it's `params.device(d) += update;`. In this sense it looks like pretty much in place. My understanding could be wrong :)